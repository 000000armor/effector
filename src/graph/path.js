//@noflow

import type {Derivable} from '../derive'
import {isDerivable} from '../derive/types'
import * as Kind from '../kind'
import type {Store} from '../store/index.h'
export type RootValue = -2
export type OpaqueValue = -1
export type NoValue = 0
export type PlainValue = 1
export type DerivedValue = 2
export type StoreValue = 3
export type StoreObjectValue = 4
export type StoreArrayValue = 5

// type RootNode = 1
// type MiddleNode = 2
// type NodeType = RootNode | MiddleNode
export type Segment =
 | [PlainValue, void]
 | [RootValue, () => any]
 | [DerivedValue, void]
 | [NoValue, void]
 | [StoreValue, void]
 | [StoreObjectValue, string]
 | [StoreArrayValue, number]
type ValueElement = Segment | [OpaqueValue, any]

type Path = ValueElement[]
type StateTree = [ValueElement, Path]

function goDeep(
 nextElement: ValueElement,
 [element, path]: StateTree,
): StateTree {
 return [nextElement, [...path, element]]
}
function flatten([element, path]: StateTree): Path {
 return [...path, element] //.filter(e => e.type !== -1)
}

function storeType(thing: any, stateTree: StateTree): StateTree[] {
 const currentTree = stateTree /* goDeep(
  {
   type: (-1: OpaqueValue),
   value: thing,
  },
  stateTree,
 )*/
 if (thing === undefined || thing === null)
  return [goDeep([(0: NoValue) /*::, undefined*/], currentTree)]
 if (isDerivable(thing)) {
  return storeTypeDerive(thing, currentTree)
 }
 if (Kind.isStore(thing)) {
  return storeTypeStore(thing, currentTree)
 }
 if (Array.isArray(thing)) {
  return storeTypeArray(thing, currentTree)
 }
 if (thing.constructor === Object) {
  return storeTypeObject(thing, currentTree)
 }
 return [goDeep([(1: PlainValue) /*::, undefined*/], stateTree)]
}
function storeTypeDerive(thing: Derivable<any>, stateTree: StateTree) {
 return storeType(
  thing.get(),
  goDeep([(2: DerivedValue) /*::, undefined*/], stateTree),
 )
}
function storeTypeStore(thing: Store<any>, stateTree: StateTree) {
 return storeType(
  thing.stateAtom.get(),
  goDeep([(3: StoreValue) /*::, undefined*/], stateTree),
 )
}
function storeTypeObject(thing: Object, stateTree: StateTree) {
 const childrens: StateTree[] = []

 for (const field of Object.keys(thing)) {
  const e = thing[field]
  const nested = storeType(e, goDeep([(4: StoreObjectValue), field], stateTree))
  childrens.push(...nested)
 }

 return childrens
}

function storeTypeArray(thing: Array<any>, stateTree: StateTree) {
 const childrens: StateTree[] = []

 for (let i = 0; i < thing.length; i++) {
  const e = thing[i]
  const nested = storeType(e, goDeep([(5: StoreArrayValue), i], stateTree))
  childrens.push(...nested)
 }

 return childrens
}

// function storeFabric<T>({defaultValue, ...opts}: {defaultValue: T}) {
//  // const isPlainStore =
//  const value: Derivable<T> = derive(() => deepUnpack(arg))
//  const proxyValue = lens({
//   get() {
//    return value.get()
//   },
//  })
//  const store = {
//   kind: Kind.STORE,
//  }
//  return store
// }
function readSegment(current: any, segment: Segment) {
 switch (segment[0]) {
  case (-2: RootValue):
   return segment[1]()
  case (0: NoValue):
   return null
  case (1: PlainValue):
   return current
  case (2: DerivedValue):
   return current.get()
  case (3: StoreValue):
   return current.stateAtom.get()
  case (4: StoreObjectValue):
   return current[segment[1]]
  case (5: StoreArrayValue):
   return current[segment[1]]
  default:
   throw new TypeError(`Wrong segment ${segment.toString()}`)
 }
}

export function readPath(path: Path): any {
 let current = null
 for (const e of path) {
  if (e[0] !== (-1: OpaqueValue)) current = readSegment(current, e)
 }
 return current
}
export function getStorePath(thing: () => any): Path[] {
 return storeType(thing(), [[(-2: RootValue), thing], []]).map(flatten)
}
