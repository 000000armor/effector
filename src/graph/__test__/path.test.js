//@flow

import {
 atom,
 derive,
 lens,
 struct,
 type Atom,
 type Derivable,
} from '../../derive'

import {createStore} from '../..'

import * as Path from '../path'

test.skip('smoke', async() => {
 expect(() => {
  Path.getStorePath(() => ({key: 'abc'}))
 }).not.toThrow()
 const store2 = {key: 'abc', atom: atom(createStore('ok'))}
 const result = Path.getStorePath(() => store2)
 console.log(result.length)
 console.log(...result[0])
 console.log(...result[1])
 console.log(Path.readPath(result[0]))
 console.log(Path.readPath(result[1]))
 expect(humanize(result)).toMatchSnapshot()
})
function humanize(list) {
 return list.map(path =>
  path.map(e => {
   switch (e[0]) {
    case (-2: Path.RootValue): //$todo
     return {type: 'root', value: e[1]()}
    case (-1: Path.OpaqueValue):
     return e[1]
    case (0: Path.NoValue):
     return 'null'
    case (1: Path.PlainValue):
     return 'plain'
    case (2: Path.DerivedValue):
     return 'derived'
    case (3: Path.StoreValue):
     return 'store'
    case (4: Path.StoreObjectValue): //$todo
     return `field ${e[1]}`
    case (5: Path.StoreArrayValue): //$todo
     return `list ${e[1]}`
    default:
     return `unknown ${e[0]}`
   }
  }),
 )
}
