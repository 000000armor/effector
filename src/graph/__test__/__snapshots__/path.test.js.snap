// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke 1`] = `
Array [
  Array [
    Object {
      "type": "root",
      "value": Object {
        "atom": Atom {
          "activeChildren": Array [],
          "equality": null,
          "id": 2,
          "kind": 21,
          "status": 2,
          "value": Object {
            "dispatch": [Function],
            "epic": [Function],
            "getState": [Function],
            "graphite": Object {
              "cmd": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
              "next": Multi {
                "data": Set {},
              },
              "seq": Seq {
                "data": Array [
                  Single {
                    "data": Compute {
                      "data": Object {
                        "reduce": [Function],
                      },
                    },
                  },
                  Multi {
                    "data": Set {},
                  },
                ],
              },
              "step": Single {
                "data": Compute {
                  "data": Object {
                    "reduce": [Function],
                  },
                },
              },
            },
            "id": "1",
            "kind": 1,
            "map": [Function],
            "on": [Function],
            "replaceReducer": [Function],
            "reset": [Function],
            "setState": [Function],
            "subscribe": [Function],
            "thru": [Function],
            "to": [Function],
            "watch": [Function],
            "withProps": [Function],
            Symbol(observable): [Function],
          },
        },
        "key": "abc",
      },
    },
    "field key",
    "plain",
  ],
  Array [
    Object {
      "type": "root",
      "value": Object {
        "atom": Atom {
          "activeChildren": Array [],
          "equality": null,
          "id": 2,
          "kind": 21,
          "status": 2,
          "value": Object {
            "dispatch": [Function],
            "epic": [Function],
            "getState": [Function],
            "graphite": Object {
              "cmd": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
              "next": Multi {
                "data": Set {},
              },
              "seq": Seq {
                "data": Array [
                  Single {
                    "data": Compute {
                      "data": Object {
                        "reduce": [Function],
                      },
                    },
                  },
                  Multi {
                    "data": Set {},
                  },
                ],
              },
              "step": Single {
                "data": Compute {
                  "data": Object {
                    "reduce": [Function],
                  },
                },
              },
            },
            "id": "1",
            "kind": 1,
            "map": [Function],
            "on": [Function],
            "replaceReducer": [Function],
            "reset": [Function],
            "setState": [Function],
            "subscribe": [Function],
            "thru": [Function],
            "to": [Function],
            "watch": [Function],
            "withProps": [Function],
            Symbol(observable): [Function],
          },
        },
        "key": "abc",
      },
    },
    "field atom",
    "derived",
    "store",
    "plain",
  ],
]
`;
